File: lib/ruby/benchmark.rb
require 'benchmark'

def variance(array)
  mean = array.sum / array.size.to_f
  sum = array.reduce(0) { |acc, cur| acc + (cur - mean) ** 2 }
  sum / (array.size - 1).to_f
end

def std_dev(array)
  Math.sqrt(variance(array))
end

def bench(run_ms)
  times = []
  result = 0
  secs = 0
  run_ns = run_ms * 1_000_000.0

  while (times.sum < run_ns && !(times.sum == 0 && times.size > 0))
    time = Benchmark.measure { result = yield }.real * 1_000_000_000
    times << time

    if run_ms > 1 && (times.sum / 1_000_000_000).round > secs
      STDERR.print '.'
      secs += 1
    end
  end
  STDERR.puts if run_ms > 1

  { times: times, result: result }
end

def format_bench(data)
  raise "no data!" if data[:times].empty?

  result = data[:result]
  times = data[:times].map { |t| t / 1_000_000.0 }

  # mean_ms,std-dev-ms,min_ms,max_ms,times,result
  "#{times.sum / times.size},#{std_dev(times)},#{times.min},#{times.max},#{times.size},#{result}"
end

File: lib/py/benchmark.py
import time
import sys

def variance(array):
    if len(array) < 2:
        return 0.0
    mean = sum(array) / len(array)
    return sum((x - mean) ** 2 for x in array) / (len(array) - 1)

def std_dev(array):
    return variance(array) ** 0.5

def bench(run_ms, func):
    times = []
    result = None
    run_ns = run_ms * 1_000_000

    while sum(times) < run_ns:
        start = time.monotonic_ns()
        result = func()
        end = time.monotonic_ns()
        elapsed = end - start
        times.append(elapsed)

        if run_ms > 1 and (sum(times) // 1_000_000_000) > (sum(times[:-1]) // 1_000_000_000):
            sys.stderr.write('.')
            sys.stderr.flush()

    if run_ms > 1:
        sys.stderr.write('\n')

    return {
        'times': times,
        'result': result
    }

def format_bench(data):
    if not data['times']:
        raise ValueError("no data!")

    result = data['result']
    times = [t / 1_000_000 for t in data['times']]  # convert to milliseconds

    # mean_ms,std-dev-ms,min_ms,max_ms,times,result
    return f"{sum(times) / len(times)},{std_dev(times)},{min(times)},{max(times)},{len(times)},{result}"